### ISSUE TITLE
Add support for hexagonal architecture

### ISSUE THREAD

**Issue by ask4gilles** [TNG/ArchUnit#89]
Currently, it is only possible to add checks on a layered architecture. Add support for hexagonal architecture.

**Comment by codecholeric** [comment]
I've already thought about that, up to now I just didn't have the time to add more predefined architectures. Do you have any suggestions, how the API should look like in code?
This should probably be discussed a little to make sure everybody has the same understanding of how this should look like, and how this should be specified. 
In particular to find a good balance between guidance and adjustability to custom projects.

**Comment by spanierm** [comment]
@ask4gilles You can have a look at https://github.com/spanierm/archunit-junit5-kotlin. We use those tests to exactly check the hexagonal architecture in our projects.

@codecholeric Totally true. If the tests I added in the above project sound reasonable for a broader audience, we can think about integrating them in ArchUnit. Up to now, I was not sure whether those tests are good enough for a broader audience. If that is the case, I volunteer to help to integrate those.

**Comment by ask4gilles** [comment]
@spanierm thanks, I'll look it up!

**Comment by codecholeric** [comment]
@spanierm Maybe we can sit down together at some Techday and check it out :wink: We should make it an API like `layeredArchitecture()` though, I think. Where you can specify something like

```
onionArchitecture()
    .domainModel("..domain.model..")
    .domainServices("..domain.services..")
    .adapters("..adapters.(*)..")
    ...
```

Just don't know how flexible it has to be to cover most use cases people have...

PS: I really like "onion architecture" better than "hexagonal architecture", especially since so many people are confused why it's six borders :wink:

**Comment by spanierm** [comment]
@codecholeric Exactly, I also find all terms very confusing and chose explicitly "onion" instead of "hexagonal". To be honest, for me it is the same. (Let me know if I am wrong ;)).

I take a note and will dig into the source code. If time allows it, I would love to contribute to ArchUnit and maybe the next Techday is good starting point.

**Comment by fdw** [comment]
I'd be interested in that as well. As always, I'm short on time to implement stuff, but I'll gladly be a sounding board.

**Comment by JonasHavers** [comment]
Since I worked with both hexagonal and onion architectures and wrote rules for them with ArchUnit, I wonder what is missing and what you expect from the API, @fdw @spanierm @ask4gilles?

Would you like to have predefined rules that you can use for these architectures so you do not have to reimplement them for every project?

The most important rule that comes to my mind is the layer dependency rule. However, I could simply write the tests with the `layeredArchitecture()`. Since the layers are not carved in stone and parts of applications like domain services are optional, I wonder what a general purpose API might look like. Do you have any ideas?

For example, that's my layer declaration for the hexagonal architecture:
```
Architectures
    .layeredArchitecture()
    .layer(layers.domainLayer.name).definedBy(layers.domainLayer.packageName)
    .layer(layers.applicationLayer.name).definedBy(layers.applicationLayer.packageName)
    .layer(layers.adaptersLayer.name).definedBy(layers.adaptersLayer.packageName)
    .layer(layers.adaptersLayer.primary.name).definedBy(layers.adaptersLayer.primary.packageName)
    .layer(layers.adaptersLayer.secondary.name).definedBy(layers.adaptersLayer.secondary.packageName)
```
Because I like to define rules for the entire adapter layer as well as individual rules for the primary (active) and secondary (passive) part of it, there is a certain differentiation here.

Another rule I commonly use checks whether domain classes only depend on other domain classes or classes from the Java (or Kotlin) standard library. For me that's another way of saying "don't use frameworks inside the domain layer". Do you want to apply such rules by default?

**Comment by fdw** [comment]
What I'd like from ArchUnit is a simple way to create rules for a common architecture. Of course I can already do everything I need, but maybe there could be a simpler API (like you suggest). Also, having a predefined ruleset makes sure I don't accidentaly forget something.

Finally, predefined rules are a way to standardize architectures. If some new developer is just starting out, it's a lot harder for them to define the architecture and the ArchUnit rules. If we can offer them something pre-built, it makes their lives easier while teaching best practices.

As for rules like "no frameworks inside the domain layer", I personally wouldn't want that as default, but maybe it could be tacked on easily?

**Comment by spanierm** [comment]
I agree with @fdw. It is simply more convenient to have a builder, e.g. as suggested by @codecholeric above. But I like the idea to use the `LayeredArchitecture` to implement the `OnionArchitecture` and started with it today :) I let you know once there is a first working version for testing.

**Comment by codecholeric** [comment]
In the end it might be just wording. I'm sure everything is already expressible more or less concise. But I would rather have a terminology that expresses "onion architecture" instead of just "layered architecture". Beneath the surface, this will probably just do something very similar to `layeredArchitecture()` :wink: 
But to express the concepts of an onion architecture instead of just assembling a lose set of rules will probably help to clarify (and standardize) the architecture. 
If it really works (i.e. how flexible it must be to cover all the different perspectives people have on an onion architecture) we need to see.

**Comment by spanierm** [comment]
I created a pull request #174 with a first version. Let me know what you think so that I can change it accordingly :)

**Comment by codecholeric** [comment]
For now I'll consider this issue solved via #174 :smiley:
